orderbook/src/account_balances.rs

```diff
@@ -114,6 +114,62 @@ impl Web3BalanceFetcher {
         }
         Ok(())
     }
+
+    async fn can_transfer_call(&self, token: H160, from: H160, amount: U256) -> bool {
+        let instance = ERC20::at(&self.web3, token);
+        let calldata = instance
+            .transfer_from(from, self.settlement_contract, amount)
+            .tx
+            .data
+            .unwrap();
+        let call_request = CallRequest {
+            from: Some(self.allowance_manager),
+            to: Some(token),
+            data: Some(calldata),
+            ..Default::default()
+        };
+        let block = Some(BlockId::Number(BlockNumber::Latest));
+        let response = self.web3.eth().call(call_request, block).await;
+        response
+            .map(|bytes| is_empty_or_truthy(bytes.0.as_slice()))
+            .unwrap_or(false)
+    }
+
+    async fn can_transfer_trace(&self, token: H160, from: H160, amount: U256) -> Result<bool> {
+        let instance = ERC20::at(&self.web3, token);
+
+        // A random account that doesn't hold any balances. Don't use the zero address as that is
+        // commonly checked against in ERC20 implementations.
+        // In addition to just transferring the balance in like can_transfer_call does, we also
+        // attempt to transfer the same balance away again. This fails for example if the token
+        // takes a fee on transfer as the full amount would be no longer be available.
+        // We use a custom recipient instead of the settlement contract because the latter might
+        // already hold some balance in the token in which case we might make up for an outgoing
+        // fee.
+        let recipient = H160::from_low_u64_le(1);
+
+        let calldata = instance
+            .transfer_from(from, recipient, amount)
+            .tx
+            .data
+            .unwrap();
+        let transfer_in = CallRequest {
+            from: Some(self.allowance_manager),
+            to: Some(token),
+            data: Some(calldata),
+            ..Default::default()
+        };
+
+        let calldata = instance.transfer(from, amount).tx.data.unwrap();
+        let transfer_out = CallRequest {
+            from: Some(recipient),
+            to: Some(token),
+            data: Some(calldata),
+            ..Default::default()
+        };
+
+        crate::trace_many::trace_many(vec![transfer_in, transfer_out], &self.web3).await
```

Comment:
Do we also need to check that the balance remains unchanged? This system can be tricked by sending in some deflationary token to the settlement contract just so this check passes.

I don't think it necessarily needs to be added here, I think this PR as-is is a great start! Just something to consider as a future improvement (if it is possible with `trace_callMany` to do get some output bytes from a call after the first two calls).

Reply:
I think, as it is now, it will only catch deflationary tokens for which the settlement contract either has no balance or a very small amount of balance.

In general, I think the assumption is fine, but in the future we can further assert that the settlement contract's balance before and after the two transfers stays the same so that the check continues to catch deflationary tokens for which the settlement contract has some balance in.

The reason for suggesting this is:
- Deflationary token admin adds settlement contract to its exemption list
- Settlement contract trades some deflationary tokens and accrues fees in the token
- Deflationary token admin removes settlement contract from exemption list
- Some trades would be allowed because of the remaining buffers in the settlement contract
- Trades with these tokens drain the settlement contract buffer for these tokens and takes fees.

Reply:
Verifying that the random account balance is the transferred out amount would also work.